pkg/controller/gitopsconfig/controller.go

```diff
@@ -389,79 +392,110 @@ func (r *Reconciler) manageDeletion(instance *gitopsv1alpha1.GitOpsConfig) (reco
-				log.Error(err, "unable to create update instance to remove finalizers", "instanceName", instance.Name)
-				return reconcile.Result{}, xerrors.Errorf("unable to create update instance %q to remove finalizers: %w", instance.Name, err)
+		log.Info("GitOpsConfig finalizer found one job", "instance", instance.Name, "podStatus", status)
+		safeReasons := []string{"ErrImagePull", "ImagePullBackOff", "InvalidImageName"}
+		if status != nil && status.Waiting != nil && containsString(safeReasons, status.Waiting.Reason) {
+			err = r.client.Delete(context.TODO(), &jobs[0], client.PropagationPolicy(metav1.DeletePropagationBackground))
+			if err != nil {
+				log.Error(err, "GitOpsConfig finalizer unable to delete job", "instance", instance.Name, "job", jobs[0].Name)
+				return reconcile.Result{}, xerrors.Errorf("GitOpsConfig finalizer unable to delete job %q for %q: %w", jobs[0].Name, instance.Name, err)
 			}
-			return reconcile.Result{}, nil
+			log.Info("GitOpsConfig finalizer deleted stuck job", "instance", instance.Name, "job", jobs[0].Name)
+			return r.removeFinalizer(context.TODO(), instance)
 		}
-		log.Info("Launching delete job for instance", "instance", instance.GetName())
-		_, err = r.createJob("delete", instance)
-		if err != nil {
-			log.Error(err, "unable to create deletion job for instance", "instanceName", instance.Name)
-			return reconcile.Result{}, xerrors.Errorf("unable to create deletion job for instance %q: %w", instance.Name, err)
+	}
+
+	// If a delete job exists, we wait (Requeue) until we can see that it
+	// completed successfully, then we remove the finalizer and we're done.
+	deleters := []batchv1.Job{}
+	for _, j := range jobs {
+		if j.Labels != nil && j.Labels["action"] == "delete" {
+			deleters = append(deleters, j)
 		}
-		//we return because we need to wait for the job to stop
-		return reconcile.Result{
-			Requeue:      true,
-			RequeueAfter: time.Second * 5,
-		}, nil
-	}
-	//There should be only one pending job
-	job := applicableJobList[0]
-	if job.Status.Succeeded > 0 {
-		instance.ObjectMeta.Finalizers = removeString(instance.ObjectMeta.Finalizers, tagFinalizer)
-		if err := r.client.Update(context.TODO(), instance); err != nil {
-			log.Error(err, "unable to create update instance to remove finalizers", "instanceName", instance.Name)
-			return reconcile.Result{}, xerrors.Errorf("unable to create update instance %q to remove finalizers: %w", instance.Name, err)
+	}
+	if len(deleters) > 0 {
+		if len(deleters) > 1 {
+			log.Error(nil, "too many delete jobs found (expected 1)", "n", len(deleters), "instance", instance.Name)
+			// TODO: should we return here, or try to still do something sensible for user?
 		}
-		return reconcile.Result{}, nil
+		done := deleters[0].Status.Succeeded > 0
+		if !done {
+			//if it's not succeeded we wait for 5 seconds
+			//TODO add logic to stop at a certain point ... or not ...
+			//TODO add exponential backoff, possibly like in CronJob
```

Comment:
Out of curiosity, in most comments there is space after `//`. Is there any specific reason why in few places it's missing? I'll also mark other code lines.